components/concept/engine/src/main/java/mozilla/components/concept/engine/InputResultDetail.kt

```diff
@@ -0,0 +1,267 @@
+ * @see [InputResultDetail.canOverscrollLeft]
+ */
+@VisibleForTesting
+internal const val OVERSCROLL_DIRECTIONS_HORIZONTAL = 1 shl 0
+
+/**
+ * The website content can be overscrolled vertically.
+ *
+ * @see [InputResultDetail.canOverscrollTop]
+ * @see [InputResultDetail.canOverscrollBottom]
+ */
+@VisibleForTesting
+internal const val OVERSCROLL_DIRECTIONS_VERTICAL = 1 shl 1
+
+/**
+ * All data about how a touch will be handled by the browser.
+ * - whether the event is used for panning/zooming by the browser / by the website or will be ignored.
+ * - whether the event can scroll the page and in what direction.
+ * - whether the event can overscroll the page and in what direction.
+ *
+ * Data wrapped by this class is safe to use after first calling [update] to update the default values.
+ *
+ * This class can be reused by calling [reset] to return all data values to it's default value.
+ */
+@Suppress("TooManyFunctions")
+class InputResultDetail {
+    /**
+     * Indicates who will use the current [MotionEvent].
+     *
+     * @see INPUT_UNHANDLED
+     * @see INPUT_HANDLED
+     * @see INPUT_HANDLED_CONTENT
+     */
+    @VisibleForTesting
+    internal var inputResult: Int = INPUT_UNHANDLED
+
+    /**
+     * Bitwise ORed value of the directions the page can be scrolled to.
+     *
+     * This is the same as GeckoView's [org.mozilla.geckoview.PanZoomController.ScrollableDirections].
+     */
+    @VisibleForTesting
+    internal var scrollDirections: Int = SCROLL_DIRECTIONS_NONE
+
+    /**
+     * Bitwise ORed value of the directions the page can be overscrolled to.
+     *
+     * This is the same as GeckoView's [org.mozilla.geckoview.PanZoomController.OverscrollDirections].
+     *
+     * The default value is [OVERSCROLL_DIRECTIONS_VERTICAL] and not [OVERSCROLL_DIRECTIONS_NONE] since
+     * there are cases in which this class can be used before valid values being set and it helps more to have
+     * overscroll vertically allowed and then stop depending on the values with which this class is updated
+     * rather than start with a disabled overscroll functionality for the current gesture.
+     */
+    @VisibleForTesting
+    internal var overscrollDirections: Int = OVERSCROLL_DIRECTIONS_VERTICAL
+
+    /**
+     * The only way to update the values contained by this class.
+     * Only after calling this method the various getters are safe to use.
+     *
+     * This method will filter out unexpected values to ensure data consistency.
+     */
+    fun update(
+        inputResult: Int? = this.inputResult,
+        scrollDirections: Int? = this.scrollDirections,
+        overscrollDirections: Int? = this.overscrollDirections
+    ) {
+        // The range check automatically checks for null but doesn't yet have a contract to say so.
+        // As such it it safe to use the not-null assertion operator.
+
+        if (inputResult in INPUT_UNHANDLED..INPUT_HANDLED_CONTENT) {
+            this.inputResult = inputResult!!
```

Comment:
As we are not handling null values, should we change the types to non null and lets the users handle the null-ability?

Reply:
The range check would automatically filter out null values.
Having this check here means a cleaner api for users and I'd say controlling what data gets stored is really the responsibility of this class and it's also nice to have it in just one place.